src/components/tree-view.tsx

```diff
@@ -0,0 +1,151 @@
+import { RightArrowIcon } from 'src/components/icons/right-arrow';
+import { cx } from 'src/lib/cx';
+
+type ExpanderValues = 'expanded' | 'collapsed';
+
+export type TreeNode = {
+	id: string;
+	label: string;
+	checked: boolean;
+	indeterminate?: boolean;
+	defaultExpanded?: boolean;
+	customExpanderOptions?: {
+		label: string;
+		value: ExpanderValues;
+	}[];
+	children?: TreeNode[];
+	type?: 'folder';
+};
+
+const updateTree = ( nodes: TreeNode[], targetId: string, checked: boolean ): TreeNode[] => {
+	return nodes.map( ( node ) => {
+		if ( node.id === targetId ) {
+			return {
+				...node,
+				checked,
+				indeterminate: false,
+				children: node.children ? updateAllChildren( node.children, checked ) : undefined,
+			};
+		} else if ( node.children ) {
+			const updatedChildren = updateTree( node.children, targetId, checked );
+			const { checked: parentChecked, indeterminate } =
+				getParentStateFromChildren( updatedChildren );
+			return {
+				...node,
+				checked: parentChecked,
+				indeterminate,
+				children: updatedChildren,
+			};
+		}
+		return node;
+	} );
+};
+
+const updateAllChildren = ( nodes: TreeNode[], checked: boolean ): TreeNode[] => {
+	return nodes.map( ( node ) => ( {
+		...node,
+		checked,
+		indeterminate: false,
+		children: node.children ? updateAllChildren( node.children, checked ) : undefined,
+	} ) );
+};
+
+const getParentStateFromChildren = (
+	children: TreeNode[]
+): { checked: boolean; indeterminate: boolean } => {
+	const total = children.length;
+	const checkedCount = children.filter( ( c ) => c.checked ).length;
+	const indeterminateCount = children.filter( ( c ) => c.indeterminate ).length;
+
+	if ( checkedCount === total ) {
+		return { checked: true, indeterminate: false };
+	}
+
+	if ( checkedCount === 0 && indeterminateCount === 0 ) {
+		return { checked: false, indeterminate: false };
+	}
+
+	return { checked: false, indeterminate: true };
+};
+
+const TreeItem = ( {
+	node,
+	onToggle,
+	level,
+}: {
+	node: TreeNode;
+	onToggle: ( id: string, checked: boolean ) => void;
+	level: number;
+} ) => {
+	const [ expanded, setExpanded ] = useState(
+		node.defaultExpanded !== undefined ? node.defaultExpanded : true
+	);
+
+	return (
+		<div className={ cx( `treeItemLevel${ level }` ) }>
+			<div className="flex items-center py-2">
+				<label className="flex items-center">
+					<CheckboxControl
+						checked={ node.checked }
+						indeterminate={ node.indeterminate }
+						onChange={ ( checked: boolean ) => onToggle( node.id, checked ) }
+						__nextHasNoMarginBottom
+					/>
+					{ node.type === 'folder' && <FolderIcon /> }
+					<span>{ node.label }</span>
+				</label>
+				{ node.children && ! node.customExpanderOptions && (
+					<button onClick={ () => setExpanded( ! expanded ) }>
+						<div className={ expanded ? 'rotate-90' : '' }>
+							<RightArrowIcon width={ 16 } />
+						</div>
+					</button>
+				) }
+
+				{ node.customExpanderOptions && (
```

Comment:
I still think this implementation has some issues.
For example, if multiple "nodes" have `customExpanderOptions` this component will render multiple `customExpanderOptions` on top of each other.
I would still prefer the `TreeView` component to expose a prop that allows this behavior, but leaving the logic and UI responsibility to the parent component. Trying to build everything inside this component will make it harder to read, understand and implement.

Another approach could be to make this component specifically to this use case, and remove all of this abstraction we are trying to create.

Reply:
Agree. I suggest having two components. One for the main checkbox with the dropdown and simplify the logic of the TreeView to just render the folders structure and nothing else.

Reply:
@nightnei I have made a quick example with how my approach would look like:
https://github.com/Automattic/studio/pull/1476